physlr/physlr.py

```diff
@@ -1021,107 +1020,118 @@ def assign_read_molecule(mxs, moltomxs, mol_counts, bx):
         return "_" + str(random.choice(max_hits)), 0, 1
 
     @staticmethod
-    def determine_molecules_biconnected_components(g, u):
-        "Separate bi-connected components."
-        cut_vertices = set(nx.articulation_points(g.subgraph(g.neighbors(u))))
-        components = list(nx.connected_components(g.subgraph(set(g.neighbors(u)) - cut_vertices)))
+    def detect_communities_biconnected_components(g, node_set):
+        """Separate bi-connected components. Return components."""
+        cut_vertices = set(nx.articulation_points(g.subgraph(node_set)))
+        components = list(nx.connected_components(g.subgraph(node_set - cut_vertices)))
         components.sort(key=len, reverse=True)
-        return u, {v: i for i, vs in enumerate(components) if len(vs) > 1 for v in vs}
-
-    @staticmethod
-    def determine_molecules_k_clique_communities(g, u):
-        "Apply k-clique community detection algorithm after extracting bi-connected components."
-        cut_vertices = set(nx.articulation_points(g.subgraph(g.neighbors(u))))
-        components = list(nx.connected_components(g.subgraph(set(g.neighbors(u)) - cut_vertices)))
-        components.sort(key=len, reverse=True)
-        communities = []
-        for comp in components:
-            if len(comp) > 1:
-                communities += list(nxcommunity.k_clique_communities(g.subgraph(comp), 3))
-        return u, {v: i for i, vs in enumerate(communities) if len(vs) > 1 for v in vs}
+        return components
 
     @staticmethod
-    def determine_molecules_louvain(g, u):
-        "Apply louvain community detection algorithm after extracting bi-connected components."
-        import community as louvain
+    def detect_communities_k_clique(g, node_set, k=3):
+        """Apply k-clique community detection. Return communities."""
+        from networkx.algorithms import community as nxcommunity
 
-        cut_vertices = set(nx.articulation_points(g.subgraph(g.neighbors(u))))
-        components = list(nx.connected_components(g.subgraph(set(g.neighbors(u)) - cut_vertices)))
-        components.sort(key=len, reverse=True)
-        communities = []
-        for comp in components:
-            if len(comp) > 1:
-                partition = louvain.best_partition(g.subgraph(comp))
-                for com in set(partition.values()):
-                    list_nodes = [nodes for nodes in partition.keys() if partition[nodes] == com]
-                    if len(list_nodes) > 1:
-                        communities.append(list_nodes)
-        return u, {v: i for i, vs in enumerate(communities) if len(vs) > 1 for v in vs}
+        if len(node_set) > 1:
```

Comment:
```suggestion
```
I don't think this test is needed.

Reply:
If we remove these lines @sjackman we'll be calling the functions for single nodes... I added this line to avoid running the algorithms for single nodes. As we have partition might have single nodes.

Reply:
Since there are no k3 cliques if you have a single node or edge, the results should be identical without this line. If the results are identical, remove this line.

Reply:
The results will be identical but I am worrying about calling k-cliques many times for nothing... and we can solve this issue just by checking a condition. So should we stick into summarizing the code or having less function calls @sjackman ?

Reply:
Calling k-cliques with a single vertex or edge will exit very quickly. There's no need to optimize for this case. You can time it and test it empirically if you're concerned.

Reply:
Just tried it to learn how to work with `timeit` :D
So here are the results:
```
>>>setup = '''
from networkx.algorithms import community as nxcommunity
import networkx as nx
node_set={1,2}
g=nx.Graph()
g.add_node(1)
g.add_node(2)
'''
>>>print(min(timeit.Timer('if len(node_set)>1: pass', setup=setup).repeat(7, 1000000)))
0.002583329100161791
>>>print(min(timeit.Timer('nxcommunity.k_clique_communities(g,3)', setup=setup).repeat(7, 1000000)))
0.04064908204600215
```
So the `if` condition is >15x faster but yep considering the calls for normal subgraphs, it's does not seem considerable.